### ISSUE TITLE
[sklearn#33737] Add `MoveTo` transformer to move `X` to a different array namespace/device

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#380]
**Mirrored from [scikit-learn#33737](https://github.com/scikit-learn/scikit-learn/issues/33737)**

Original title: Add `MoveTo` transformer to move `X` to a different array namespace/device
Original labels: Needs Decision, Array API
Created: 2026-04-13T11:57:59Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

This issue describes a new transformer `MoveTo` that can be used to move `X` to another array namespace and device. The goal of the transformer is to move the input array `X` to a specified array namespace (numpy, torch, cupy) and device (CPU, CUDA, MPS, etc). The transformer can not move `y` or `sample_weight` but that is Ok, because of https://github.com/scikit-learn/scikit-learn/issues/28668.

The goal is to [enable a `Pipeline` like the following](https://github.com/scikit-learn/scikit-learn/issues/31274#issuecomment-2884368423) when using the experimental support for array API (and other use-cases where processing has to move from one namespace/device to another):

```python
df = pd.read_parquet("data_file.parquet")
X = df[categorical_feature_names]  # categorical or str/object dtype
y = df[target_column_name]  # continuous value to regress.

pipeline = make_pipeline(
   TargetEncoder(),  # works on categorical X but also needs y, hence CPU for both
   MoveTo(namespace=torch, device="cuda:0"),
   Ridge(),  # only numerical inputs: faster to do it all on GPU, y follows X => y_pred on GPU.
)

cv_results = cross_validate(pipeline, X, y, scoring="r2")
```

Currently you can make the example above work by replacing `MoveTo` with `FunctionTransformer(partial(torch.asarray, device="cuda:0"))`. However it is difficult for the average user to discover this particular `FunctionTransformer` incantation and looking towards the future where we want to offer a function that can "translate" a whole pipeline from running on one namespace/device to another (e.g. train on GPU, then deploy on CPU) having an explicitly named step that can be specially treated is useful.

xref https://github.com/scikit-learn/scikit-learn/issues/33444

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [16](https://github.com/godofecht/flow-scikit/issues/16) in the flow-scikit repository. Further library-specific triage and implementation will happen there.

**Comment by godofecht** [comment]
Moved to [17](https://github.com/godofecht/flow-scikit/issues/17) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
